torch/_export/__init__.py

```diff
@@ -0,0 +1,228 @@
+import contextlib
+import copy
+from typing import Callable, Tuple, Generator, Dict
+
+import torch
+import torch._dynamo as torchdynamo
+from torch._decomp import core_aten_decompositions
+from torch._dispatch.python import enable_python_dispatcher
+from torch.nn.utils import stateless
+from torch.utils import _pytree as pytree
+
+from torch._functorch.aot_autograd import (
+    AOTConfig,
+    create_aot_dispatcher_function,
+    default_partition,
+    run_functionalized_fw_and_collect_metadata,
+)
+
+from torch.fx.experimental.proxy_tensor import (
+    get_proxy_slot,
+    get_torch_dispatch_modes,
+    has_proxy_slot,
+    make_fx,
+    ProxyTorchDispatchMode,
+    set_proxy_slot,
+)
+
+from torch._functorch.eager_transforms import _unwrap_all_tensors_from_functional
+
+from .workflow import ExportedProgram
+
+CORE_ATEN_DECOMPOSITIONS_TABLE = core_aten_decompositions()
+
+
+@contextlib.contextmanager
+def _using_dynamo_config(**kwargs):
+    prev_configs : Dict[str, bool] = {}
+    for key in kwargs:
+        assert hasattr(torchdynamo.config, key)
+        prev_configs[key] = getattr(torchdynamo.config, key)
+
+    for key in kwargs:
+        setattr(torchdynamo.config, key, kwargs[key])
+    try:
+        yield
+    finally:
+        for key, value in prev_configs.items():
+            setattr(torchdynamo.config, key, value)
+
+def aot_capture(mod, flat_args):
+    """
+    A wrapper around aot_autograd() to mix AOT Autograd + torch.export.
+    Some assumptions were made about the AOT Autograd internal:
+    1. The functionalization metadata format.
+    2. Calling convention of returned forward graph.
+    3. make_fx() internal proxy storage.
+
+    In the current context we're just experimenting the idea so it's possible things
+    could break. For the next step we should find a way to upstream something reasonable.
+    """
+    param_list = [
+        *mod.named_parameters(remove_duplicate=False),
+        *mod.named_buffers(remove_duplicate=False),
+    ]
+    params = dict(param_list)
+    params_flat, params_spec = pytree.tree_flatten(params)
+    params_len = len(params_flat)
```

Comment:
This is just taking code from aot_autograd.py and reimplementing it, why?

Reply:
Oh this is actually coming from Zhengxu's experiment on using AOTAutograd instead of make_fx. Since this change handles our use cases pretty well, this can serve as placeholder for our tests to pass as baseline. Soon, this will be replaced with functionalization (used in AOTAutograd) + make_fx.

Reply:
per @SherlockNoMad's advice we don't want to reuse aotautograd for now.

Reply:
Pull into a util

Reply:
@tugsbayasgalan Shall we pull into a util so avoid forking code like Voz said?